inference-engine/include/ie_compound_blob.h

```diff
@@ -349,4 +349,64 @@ class INFERENCE_ENGINE_API_CLASS(I420Blob) : public CompoundBlob {
 
     Blob::Ptr createROI(const ROI& roi) const override;
 };
+
+/**
+ * @brief This class represents a blob that contains other blobs - one per batch
+ */
+class INFERENCE_ENGINE_API_CLASS(BatchedBlob) : public CompoundBlob {
+ public:
+    /**
+     * @brief A smart pointer to the BatchedBlob object
+     */
+    using Ptr = std::shared_ptr<BatchedBlob>;
+
+    /**
+     * @brief A smart pointer to the const BatchedBlob object
+     */
+    using CPtr = std::shared_ptr<const BatchedBlob>;
+
+    /**
+     * @brief Constructs a batched blob from a vector of blobs
+     * @details All passed blobs should meet following requirements:
+     * - all blobs have equal tensor descriptors,
+     * - blobs layouts should be one of: NCHW, NHWC, NCDHW, NDHWC, NC, CN, C, CHW
+     * - batch dimensions should be equal to 1 or not defined (C, CHW).
+     * Resulting blob's tensor descriptor is constructed using tensor descriptors
+     * of passed blobs by setting batch dimension to blobs.size()
+     *
+     * @param blobs A vector of blobs that is copied to this object
+     */
+    explicit BatchedBlob(std::vector<Blob::Ptr> blobs);
+
+    /**
+     * @brief Constructs a batched blob from a vector of blobs with specified
+     * tensor descriptor
+     * @details It is caller's responsibility to validate the correctness of blobs
+     * and tensor descriptor
+     *
+     * @param tensorDesc Defines the layout and dims of the blob
+     * @param blobs A vector of blobs that is copied to this object
+     */
+    BatchedBlob(std::vector<Blob::Ptr> blobs, const TensorDesc& tensorDesc);
+
+    /**
+     * @brief A virtual destructor
+     */
+    virtual ~BatchedBlob() = default;
+
+    /**
+     * @brief Copy constructor
+     */
+    BatchedBlob(const BatchedBlob& other);
+
+    /**
+     * @brief Move constructor
+     */
+    BatchedBlob(BatchedBlob&& other);
+
+    /**
+     * @brief Assignment operator
+     */
+    BatchedBlob& operator=(BatchedBlob other);
```

Comment:
Why by value?

Reply:
It is just the copy-and-swap idiom

Reply:
As far as I know it's not recommended due to possible performance penalty in case of lvalue argument

Reply:
I just tried to google uses of copy and swap idiom in modern C++ and saw paper discussing why it's not preferable way since C++11 (check [this](http://scottmeyers.blogspot.com/2014/06/the-drawbacks-of-implementing-move.html) out). Anyway even there argument is passed via r-value reference.

Reply:
[Using the swap idiom is not perfectly performant, but is generally not overly expensive and correct.](http://www.vollmann.ch/en/blog/implementing-move-assignment-variations-in-c++.html)

Anyway, "delay of resource release" is not an option here: https://coliru.stacked-crooked.com/a/729814a3a4542cf2

Anyway, I would follow the "rule-of-zero", when base classes will follow this rule.

Reply:
Thanks for your experiments!

From them I see that swap idiom is less performant:

Classic case:

```
Copy assignment...
Allocating resource: three
A::A()
A::operator=(const A&)
Copy-assigning resource: three=one
Releasing resource: three
Copy-assigned resource: one-copy
```
vs

Copy-And-Swap
```
Copy assignment...
Allocating resource: three
B::B()
Copying resource: one-copy
B::B(const B&)
B::operator=(B)
Moving resource: three
Move-assigning resource: =one-copy
Releasing resource: 
Move-assigned resource: one-copy
Move-assigning resource: =three
Releasing resource: 
Move-assigned resource: three
Releasing resource: <empty>
B::~B()
Releasing resource: three
```

And as I understand, it's becoming worse with growing vector size, because in case of l-value reference compiler is able to reuse space that taken by the object

Consider the example:

```cpp
A& A::operator=(A other) { // here before call we do copy of all elements in others.buffer
    // here we kind of optimize
    std::swap(this.buffer, other.buffer);
    return *this;
}
```
vs classic approach

```cpp
A& A::operator=(A const& other) { // no copy
    // there compiler knows exact sizes of vector so have more context and can optimize memory allocation and copies
    this.buffer = other.buffer;
}
```

So I agree with the statement that copy-and-swap is pessimization in terms of optimization context and not recommended for usage

Regarding

> I would prefer to follow the "rule-of-zero", when base classes will follow this rule.

I agree that rule-of-zero simplifies situation a bit. I'm not sure if it can be used here or not, but even if not, I think we have better option than swap-and-copy

Reply:
> Thanks for your experiments!
You are welcome. :)
> So I agree with the statement that copy-and-swap is pessimization in terms of optimization context and not recommended for usage
Agree, if performance is more important than code clearance and exception safety for copy assignment.

Sorry, in when I see performance arguments, I cannot resist from pointing to that: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp#L825

Reply:
> Agree, if performance is more important than code clearance and exception safety for copy assignment.

To me version with copy and swap looks less clear actually, that's why I started the discussion. I understand your point that it was popular idiom before C++11, but looks like it's no more an idiomatic code. But this opinion is rather subjective because I started to learn C++ after C++11 appeared, so I never used and heard about this idiom. In the context of C++11 usage and C++17 coming in next year to our project (at least I hope) I think there is no reason to introduce probably outdated idioms.

> Sorry, it is hard to resist from pointing to that in case of performance discussions.

Sorry, I'm not sure I got your point. Do you mean that there we have insertion at the beginning of the vector, so a lot of allocations and copies are happening? If so, probably it's the issue indeed, maybe it's reasonable (I don't know the context) - but anyway I don't see a way it affects code you added in your PR. If you feel it's the place requiring refactoring I think the best option to raise an improvement ticket about that and share a solution in a separate PR.

Anyway, I got your idea, proposed my opinion, but I don't insist. Thanks.

Reply:
Ok.

*Usually* (in 90% cases), swap-based assignments are:

1. Easy to implement - e.g. no need to handle self-assignment case.
2. Exception safe - unfortunately, not in this case due to `std::vector` in C++11
3. Less risky to introduce an issue - e.g. no need to explicitly handle resources lifetime inside assignment operator.
4. Reduce code duplication -> maintainability

The cost of these benefit is the some small performance impact by extra move operations (as mentioned in articles referred in this discussion) - which usually eliminated by compiler optimization.

For sure, in case of performance critical code, this approach _might_ be ineffective - especially, if class members and base classes not properly defined.\
But in _most_ cases, benefits of the approach noted above exceed benefits of micro-optimizations.

I referred the existing IE code just to demonstrate that micro-optimizations is not the good point for discussion when there are many other areas for performance improvements.

PS. Simplified exception-safe assignment example:
```cpp
A& operator=(const A& other) {
    if(this == &other) return *this;
    A tmp{other}; // copy constructor can throw from fields copy operations
    std::swap(*this, tmp); // assuming explicitly defined: A& A::operator=(A&&) noexcept; or std::swap(A&, A&) noexcept;
    return *this;
}
```

Reply:
N/A due to applied `rule-of-zero`